head_test.go

```diff
@@ -92,72 +92,103 @@ func TestAmendDatapointCausesError(t *testing.T) {
-	if err != nil {
-		t.Fatalf("Error creating head block: %s", err)
-	}
+	require.NoError(t, err, "Error creating head block")
 
 	app := hb.Appender()
 	_, err = app.Add(labels.Labels{}, 0, 0)
-	if err != nil {
-		t.Fatalf("Failed to add sample: %s", err)
-	}
-	if err = app.Commit(); err != nil {
-		t.Fatalf("Unexpected error committing appender: %s", err)
-	}
+	require.NoError(t, err, "Failed to add sample")
+	require.NoError(t, app.Commit(), "Unexpected error committing appender")
 
 	app = hb.Appender()
 	_, err = app.Add(labels.Labels{}, 0, 1)
-	if err != ErrAmendSample {
-		t.Fatalf("Expected error amending sample, got: %s", err)
-	}
+	require.Equal(t, ErrAmendSample, err)
 }
 
 func TestDuplicateNaNDatapointNoAmendError(t *testing.T) {
 	tmpdir, _ := ioutil.TempDir("", "test")
 	defer os.RemoveAll(tmpdir)
 
 	hb, err := createHeadBlock(tmpdir+"/hb", 0, nil, 0, 1000)
-	if err != nil {
-		t.Fatalf("Error creating head block: %s", err)
-	}
+	require.NoError(t, err, "Error creating head block")
 
 	app := hb.Appender()
 	_, err = app.Add(labels.Labels{}, 0, math.NaN())
-	if err != nil {
-		t.Fatalf("Failed to add sample: %s", err)
-	}
-	if err = app.Commit(); err != nil {
-		t.Fatalf("Unexpected error committing appender: %s", err)
-	}
+	require.NoError(t, err, "Failed to add sample")
+	require.NoError(t, app.Commit(), "Unexpected error committing appender")
 
 	app = hb.Appender()
 	_, err = app.Add(labels.Labels{}, 0, math.NaN())
-	if err != nil {
-		t.Fatalf("Unexpected error adding duplicate NaN sample, got: %s", err)
-	}
+	require.NoError(t, err)
 }
 
 func TestNonDuplicateNaNDatapointsCausesAmendError(t *testing.T) {
 	tmpdir, _ := ioutil.TempDir("", "test")
 	defer os.RemoveAll(tmpdir)
 
 	hb, err := createHeadBlock(tmpdir+"/hb", 0, nil, 0, 1000)
-	if err != nil {
-		t.Fatalf("Error creating head block: %s", err)
-	}
+	require.NoError(t, err, "Error creating head block")
 
 	app := hb.Appender()
 	_, err = app.Add(labels.Labels{}, 0, math.Float64frombits(0x7ff0000000000001))
-	if err != nil {
-		t.Fatalf("Failed to add sample: %s", err)
-	}
-	if err = app.Commit(); err != nil {
-		t.Fatalf("Unexpected error committing appender: %s", err)
-	}
+	require.NoError(t, err, "Failed to add sample")
+	require.NoError(t, app.Commit(), "Unexpected error committing appender")
 
 	app = hb.Appender()
 	_, err = app.Add(labels.Labels{}, 0, math.Float64frombits(0x7ff0000000000002))
-	if err != ErrAmendSample {
-		t.Fatalf("Expected error amending sample, got: %s", err)
-	}
+	require.Equal(t, ErrAmendSample, err)
+}
+
+func TestSkippingInvalidValuesInSameTxn(t *testing.T) {
+	tmpdir, _ := ioutil.TempDir("", "test")
+	defer os.RemoveAll(tmpdir)
+
+	hb, err := createHeadBlock(tmpdir+"/hb", 0, nil, 0, 1000)
+	require.NoError(t, err)
+
+	// Append AmendedValue
+	app := hb.Appender()
+	_, err = app.Add(labels.Labels{{"a", "b"}}, 0, 1)
+	require.NoError(t, err)
+	_, err = app.Add(labels.Labels{{"a", "b"}}, 0, 2)
```

Comment:
Multiple samples for the same series in the same transaction are currently indeed not verified against each other. This would need same changes in how we batch up samples in the transaction object as you said.

For now it's fine to keep it, adding a test that we are not writing anything invalid is certainly good.